File: modlab-cpp-client/modlab/src/ui/NewPackDialog.cpp
#include "NewPackDialog.h"
#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QLabel>
#include <QFrame>
#include <QStyle>

NewPackDialog::NewPackDialog(QWidget* parent)
    : QDialog(parent)
{
    setObjectName("NewPackDialog");
    setWindowTitle("New Instance");
    setModal(true);
    setFixedSize(370, 340);
    // Frameless rounded dialog — real border is drawn by QSS
    setWindowFlags(Qt::Dialog | Qt::FramelessWindowHint);
    setAttribute(Qt::WA_TranslucentBackground);
    buildLayout();
}

void NewPackDialog::buildLayout() {
    auto* root = new QVBoxLayout(this);
    root->setContentsMargins(0,0,0,0);
    root->setSpacing(0);

    // ---- Header ----
    auto* header = new QWidget(this);
    header->setFixedHeight(52);
    header->setStyleSheet(
        "background: qlineargradient(x1:0,y1:0,x2:0,y2:1,"
        "stop:0 rgba(56,217,245,0.06),stop:1 transparent);"
        "border-bottom: 1px solid rgba(56,217,245,0.1);"
        "border-top-left-radius: 12px; border-top-right-radius: 12px;"
    );
    auto* hLayout = new QHBoxLayout(header);
    hLayout->setContentsMargins(22,0,14,0);

    auto* titleLbl = new QLabel(header);
    titleLbl->setText("New <span style='color:#38d9f5'>Instance</span>");
    titleLbl->setTextFormat(Qt::RichText);
    titleLbl->setStyleSheet("font-family:'Exo 2'; font-weight:700; font-size:14px; color:#d8f0ff;");

    auto* closeBtn = new QPushButton("✕", header);
    closeBtn->setFixedSize(26, 22);
    closeBtn->setStyleSheet(
        "QPushButton { background: qlineargradient(x1:0,y1:0,x2:0,y2:1,"
        "stop:0 rgba(255,255,255,0.07),stop:1 rgba(0,0,0,0.1));"
        "border-top:1px solid rgba(255,255,255,0.1);"
        "border-bottom:1px solid rgba(0,0,0,0.3);"
        "border-left:1px solid rgba(255,255,255,0.05);"
        "border-right:1px solid rgba(0,0,0,0.2);"
        "border-radius:4px; color:#7ab8d4; font-size:11px; }"
        "QPushButton:hover { color:#d8f0ff; }"
    );
    connect(closeBtn, &QPushButton::clicked, this, &QDialog::reject);

    hLayout->addWidget(titleLbl);
    hLayout->addStretch();
    hLayout->addWidget(closeBtn);
    root->addWidget(header);

    // ---- Body ----
    auto* body = new QWidget(this);
    body->setStyleSheet("background: transparent;");
    auto* bodyLayout = new QVBoxLayout(body);
    bodyLayout->setContentsMargins(22, 18, 22, 18);
    bodyLayout->setSpacing(13);

    auto makeFieldLabel = [](const QString& text, QWidget* parent) {
        QLabel* lbl = new QLabel(
            "<span style='color:#00e5cc'>&#62;</span> " + text,
            parent
        );
        lbl->setTextFormat(Qt::RichText);
        lbl->setStyleSheet("font-size: 9px; letter-spacing: 3px; color: #7ab8d4;");
        return lbl;
    };

    // Pack name
    auto* nameBlock = new QWidget(body);
    auto* nameLayout = new QVBoxLayout(nameBlock);
    nameLayout->setContentsMargins(0,0,0,0);
    nameLayout->setSpacing(5);
    nameLayout->addWidget(makeFieldLabel("PACK NAME", nameBlock));
    m_nameEdit = new QLineEdit(nameBlock);
    m_nameEdit->setObjectName("FieldInput");
    m_nameEdit->setPlaceholderText("e.g. My Survival World");
    nameLayout->addWidget(m_nameEdit);
    bodyLayout->addWidget(nameBlock);

    // Version
    auto* verBlock = new QWidget(body);
    auto* verLayout = new QVBoxLayout(verBlock);
    verLayout->setContentsMargins(0,0,0,0);
    verLayout->setSpacing(5);
    verLayout->addWidget(makeFieldLabel("MINECRAFT VERSION", verBlock));
    m_versionEdit = new QLineEdit(verBlock);
    m_versionEdit->setObjectName("FieldInput");
    m_versionEdit->setPlaceholderText("e.g. 1.21.4");
    verLayout->addWidget(m_versionEdit);
    bodyLayout->addWidget(verBlock);

    // Loader selector
    auto* loaderBlock = new QWidget(body);
    auto* loaderLayout = new QVBoxLayout(loaderBlock);
    loaderLayout->setContentsMargins(0,0,0,0);
    loaderLayout->setSpacing(5);
    loaderLayout->addWidget(makeFieldLabel("MOD LOADER", loaderBlock));

    auto* loaderRow = new QHBoxLayout;
    loaderRow->setSpacing(7);

    auto makeLoaderBtn = [&](const QString& abbr, const QString& name) -> QPushButton* {
        QPushButton* btn = new QPushButton(loaderBlock);
        btn->setFixedHeight(56);
        btn->setText(QString("<span style='font-size:12px;font-weight:900;"
                             "font-family:Exo 2;letter-spacing:1px'>%1</span>"
                             "<br><span style='font-size:8px;letter-spacing:2px;"
                             "color:#7ab8d4'>%2</span>").arg(abbr, name));
        // QPushButton doesn't render HTML natively; we use a QLabel trick below instead
        return btn;
    };
    (void)makeLoaderBtn; // suppress unused warning — using labels instead

    // Use QFrame+QLabel for each loader tile (QPushButton doesn't do multiline HTML well)
    auto makeTile = [&](const QString& abbr, const QString& label, Loader loader) -> QFrame* {
        QFrame* tile = new QFrame(loaderBlock);
        tile->setFixedHeight(56);
        tile->setCursor(Qt::PointingHandCursor);
        tile->setProperty("loaderSel", false);

        auto applyTileStyle = [tile](bool selected) {
            if (selected) {
                tile->setStyleSheet(
                    "QFrame { background: qlineargradient(x1:0,y1:0,x2:0.5,y2:1,"
                    "stop:0 rgba(56,217,245,0.12),stop:1 rgba(20,80,110,0.15));"
                    "border-top:1px solid rgba(56,217,245,0.5);"
                    "border-left:1px solid rgba(56,217,245,0.22);"
                    "border-right:1px solid rgba(0,0,0,0.25);"
                    "border-bottom:1px solid rgba(0,0,0,0.45);"
                    "border-radius:6px; }"
                );
            } else {
                tile->setStyleSheet(
                    "QFrame { background: qlineargradient(x1:0,y1:0,x2:0.5,y2:1,"
                    "stop:0 rgba(10,30,52,0.7),stop:1 rgba(4,14,26,0.8));"
                    "border-top:1px solid rgba(255,255,255,0.08);"
                    "border-left:1px solid rgba(255,255,255,0.04);"
                    "border-right:1px solid rgba(0,0,0,0.25);"
                    "border-bottom:1px solid rgba(0,0,0,0.45);"
                    "border-radius:6px; }"
                    "QFrame:hover { border-top-color: rgba(56,217,245,0.25); }"
                );
            }
        };
        applyTileStyle(loader == Loader::Fabric); // Fabric selected by default

        auto* tileLayout = new QVBoxLayout(tile);
        tileLayout->setContentsMargins(0,6,0,6);
        tileLayout->setAlignment(Qt::AlignCenter);

        QLabel* abbrLbl = new QLabel(abbr, tile);
        abbrLbl->setAlignment(Qt::AlignCenter);
        abbrLbl->setStyleSheet(
            loader == Loader::Fabric
            ? "font-family:'Exo 2'; font-weight:900; font-size:12px; color:#38d9f5; letter-spacing:1px;"
            : "font-family:'Exo 2'; font-weight:900; font-size:12px; color:#7ab8d4; letter-spacing:1px;"
        );

        QLabel* nameLbl = new QLabel(label, tile);
        nameLbl->setAlignment(Qt::AlignCenter);
        nameLbl->setStyleSheet("font-size:8px; letter-spacing:2px; color:#4a7a96; text-transform:uppercase;");

        tileLayout->addWidget(abbrLbl);
        tileLayout->addWidget(nameLbl);

        // Click handler via event filter isn't needed — use mousePressEvent via installEventFilter
        // Simpler: use a transparent QPushButton overlaid, but cleanest is subclassing.
        // Here we capture with a child QPushButton that fills the tile.
        auto* clickTarget = new QPushButton(tile);
        clickTarget->setFixedSize(tile->width(), 56);
        clickTarget->move(0,0);
        clickTarget->setStyleSheet("background:transparent; border:none;");
        clickTarget->raise();

        connect(clickTarget, &QPushButton::clicked, this, [this, loader, applyTileStyle,
                abbrLbl, label]() mutable {
            selectLoader(loader);
        });
        // Store applyTileStyle ref per tile — easier to just refresh all in selectLoader
        // Store tile pointer indexed by loader for re-styling
        tile->setObjectName(QString("loader_%1").arg((int)loader));

        return tile;
    };

    m_loaderFabric  = nullptr; // using tiles instead
    QFrame* fabricTile  = makeTile("FBR", "Fabric",  Loader::Fabric);
    QFrame* forgeTile   = makeTile("FRG", "Forge",   Loader::Forge);
    QFrame* vanillaTile = makeTile("VNL", "Vanilla", Loader::Vanilla);

    loaderRow->addWidget(fabricTile);
    loaderRow->addWidget(forgeTile);
    loaderRow->addWidget(vanillaTile);
    loaderLayout->addLayout(loaderRow);
    bodyLayout->addWidget(loaderBlock);
    root->addWidget(body, 1);

    // ---- Footer ----
    auto* footer = new QWidget(this);
    footer->setFixedHeight(52);
    footer->setStyleSheet(
        "background: transparent;"
        "border-top: 1px solid rgba(56,217,245,0.1);"
        "border-bottom-left-radius: 12px; border-bottom-right-radius: 12px;"
    );
    auto* footLayout = new QHBoxLayout(footer);
    footLayout->setContentsMargins(22, 0, 22, 0);

    // Step dots
    auto* dotsRow = new QHBoxLayout;
    dotsRow->setSpacing(5);
    for (int i = 0; i < 3; ++i) {
        QLabel* dot = new QLabel(footer);
        dot->setFixedSize(6,6);
        dot->setStyleSheet(i == 0
            ? "background:#38d9f5; border-radius:3px;"
            : "background:rgba(56,217,245,0.15); border:1px solid rgba(56,217,245,0.12); border-radius:3px;"
        );
        dotsRow->addWidget(dot);
    }
    footLayout->addLayout(dotsRow);
    footLayout->addStretch();

    auto* cancelBtn = new QPushButton("Cancel", footer);
    cancelBtn->setObjectName("BtnSecondary");
    connect(cancelBtn, &QPushButton::clicked, this, &QDialog::reject);

    m_btnCreate = new QPushButton("Create", footer);
    m_btnCreate->setObjectName("BtnPrimary");
    connect(m_btnCreate, &QPushButton::clicked, this, [this]() {
        m_result = { m_nameEdit->text().trimmed(),
                     m_versionEdit->text().trimmed(),
                     m_loader };
        accept();
    });

    footLayout->addWidget(cancelBtn);
    footLayout->addSpacing(7);
    footLayout->addWidget(m_btnCreate);
    root->addWidget(footer);

    connect(m_nameEdit,    &QLineEdit::textChanged, this, &NewPackDialog::validate);
    connect(m_versionEdit, &QLineEdit::textChanged, this, &NewPackDialog::validate);
    validate();
}

void NewPackDialog::selectLoader(Loader l) {
    m_loader = l;
    // Re-style all loader tiles
    static const QMap<Loader,QString> names = {
        {Loader::Fabric,"Fabric"},{Loader::Forge,"Forge"},{Loader::Vanilla,"Vanilla"}
    };
    // Find tiles by objectName
    for (Loader tl : {Loader::Fabric, Loader::Forge, Loader::Vanilla}) {
        QFrame* tile = findChild<QFrame*>(QString("loader_%1").arg((int)tl));
        if (!tile) continue;
        bool sel = (tl == l);
        if (sel) {
            tile->setStyleSheet(
                "QFrame { background: qlineargradient(x1:0,y1:0,x2:0.5,y2:1,"
                "stop:0 rgba(56,217,245,0.12),stop:1 rgba(20,80,110,0.15));"
                "border-top:1px solid rgba(56,217,245,0.5);"
                "border-left:1px solid rgba(56,217,245,0.22);"
                "border-right:1px solid rgba(0,0,0,0.25);"
                "border-bottom:1px solid rgba(0,0,0,0.45);"
                "border-radius:6px; }"
            );
            // Update abbr label color
            QList<QLabel*> labels = tile->findChildren<QLabel*>();
            if (!labels.isEmpty())
                labels[0]->setStyleSheet(
                    "font-family:'Exo 2'; font-weight:900; font-size:12px;"
                    "color:#38d9f5; letter-spacing:1px;"
                );
        } else {
            tile->setStyleSheet(
                "QFrame { background: qlineargradient(x1:0,y1:0,x2:0.5,y2:1,"
                "stop:0 rgba(10,30,52,0.7),stop:1 rgba(4,14,26,0.8));"
                "border-top:1px solid rgba(255,255,255,0.08);"
                "border-left:1px solid rgba(255,255,255,0.04);"
                "border-right:1px solid rgba(0,0,0,0.25);"
                "border-bottom:1px solid rgba(0,0,0,0.45);"
                "border-radius:6px; }"
                "QFrame:hover { border-top-color:rgba(56,217,245,0.25); }"
            );
            QList<QLabel*> labels = tile->findChildren<QLabel*>();
            if (!labels.isEmpty())
                labels[0]->setStyleSheet(
                    "font-family:'Exo 2'; font-weight:900; font-size:12px;"
                    "color:#7ab8d4; letter-spacing:1px;"
                );
        }
    }
}

void NewPackDialog::validate() {
    bool ok = !m_nameEdit->text().trimmed().isEmpty()
           && !m_versionEdit->text().trimmed().isEmpty();
    m_btnCreate->setEnabled(ok);
    m_btnCreate->setObjectName(ok ? "BtnPrimary" : "BtnSecondary");
    m_btnCreate->style()->unpolish(m_btnCreate);
    m_btnCreate->style()->polish(m_btnCreate);
}

File: modlab/ui/new_pack_dialog.py
from dataclasses import dataclass

from PySide6.QtCore import Qt
from PySide6.QtWidgets import QDialog, QFrame, QHBoxLayout, QLabel, QLineEdit, QPushButton, QVBoxLayout, QWidget

from ..core.pack import Loader


@dataclass
class NewPackData:
    name: str = ""
    version: str = ""
    loader: Loader = Loader.FABRIC


class NewPackDialog(QDialog):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.setObjectName("NewPackDialog")
        self.setWindowTitle("New Instance")
        self.setModal(True)
        self.setFixedSize(370, 340)
        self.setWindowFlags(Qt.Dialog | Qt.FramelessWindowHint)
        self.setAttribute(Qt.WA_TranslucentBackground)
        self.loader = Loader.FABRIC
        self.result_data = NewPackData()
        self.loader_tiles: dict[Loader, QFrame] = {}
        self._build_layout()

    def _build_layout(self) -> None:
        root = QVBoxLayout(self)
        root.setContentsMargins(0, 0, 0, 0)
        root.setSpacing(0)

        header = QWidget(self)
        header.setFixedHeight(52)
        header.setStyleSheet(
            "background: qlineargradient(x1:0,y1:0,x2:0,y2:1,"
            "stop:0 rgba(56,217,245,0.06),stop:1 transparent);"
            "border-bottom: 1px solid rgba(56,217,245,0.1);"
            "border-top-left-radius: 12px; border-top-right-radius: 12px;"
        )
        header_layout = QHBoxLayout(header)
        header_layout.setContentsMargins(22, 0, 14, 0)
        title = QLabel("New <span style='color:#38d9f5'>Instance</span>", header)
        title.setStyleSheet("font-family:'Exo 2'; font-weight:700; font-size:14px; color:#d8f0ff;")
        close = QPushButton("x", header)
        close.setFixedSize(26, 22)
        close.setStyleSheet(
            "QPushButton { background: qlineargradient(x1:0,y1:0,x2:0,y2:1,"
            "stop:0 rgba(255,255,255,0.07),stop:1 rgba(0,0,0,0.1));"
            "border-top:1px solid rgba(255,255,255,0.1);"
            "border-bottom:1px solid rgba(0,0,0,0.3);"
            "border-radius:4px; color:#7ab8d4; font-size:11px; }"
            "QPushButton:hover { color:#d8f0ff; }"
        )
        close.clicked.connect(self.reject)
        header_layout.addWidget(title)
        header_layout.addStretch()
        header_layout.addWidget(close)
        root.addWidget(header)

        body = QWidget(self)
        body.setStyleSheet("background: transparent;")
        body_layout = QVBoxLayout(body)
        body_layout.setContentsMargins(22, 18, 22, 18)
        body_layout.setSpacing(13)

        self.name_edit = self._add_field(body, body_layout, "PACK NAME", "e.g. My Survival World")
        self.version_edit = self._add_field(body, body_layout, "MINECRAFT VERSION", "e.g. 1.21.4")

        loader_block = QWidget(body)
        loader_layout = QVBoxLayout(loader_block)
        loader_layout.setContentsMargins(0, 0, 0, 0)
        loader_layout.setSpacing(5)
        loader_layout.addWidget(self._field_label("MOD LOADER", loader_block))
        row = QHBoxLayout()
        row.setSpacing(7)
        for abbr, label, loader in (
            ("FBR", "Fabric", Loader.FABRIC),
            ("FRG", "Forge", Loader.FORGE),
            ("VNL", "Vanilla", Loader.VANILLA),
        ):
            tile = self._make_loader_tile(loader_block, abbr, label, loader)
            self.loader_tiles[loader] = tile
            row.addWidget(tile)
        loader_layout.addLayout(row)
        body_layout.addWidget(loader_block)
        root.addWidget(body, 1)

        footer = QWidget(self)
        footer.setFixedHeight(52)
        footer.setStyleSheet(
            "background: transparent; border-top: 1px solid rgba(56,217,245,0.1);"
            "border-bottom-left-radius: 12px; border-bottom-right-radius: 12px;"
        )
        footer_layout = QHBoxLayout(footer)
        footer_layout.setContentsMargins(22, 0, 22, 0)
        dots = QHBoxLayout()
        dots.setSpacing(5)
        for i in range(3):
            dot = QLabel(footer)
            dot.setFixedSize(6, 6)
            dot.setStyleSheet(
                "background:#38d9f5; border-radius:3px;"
                if i == 0
                else "background:rgba(56,217,245,0.15); border:1px solid rgba(56,217,245,0.12); border-radius:3px;"
            )
            dots.addWidget(dot)
        footer_layout.addLayout(dots)
        footer_layout.addStretch()

        cancel = QPushButton("Cancel", footer)
        cancel.setObjectName("BtnSecondary")
        cancel.clicked.connect(self.reject)
        self.create_button = QPushButton("Create", footer)
        self.create_button.setObjectName("BtnPrimary")
        self.create_button.clicked.connect(self._accept_result)
        footer_layout.addWidget(cancel)
        footer_layout.addSpacing(7)
        footer_layout.addWidget(self.create_button)
        root.addWidget(footer)

        self.name_edit.textChanged.connect(self._validate)
        self.version_edit.textChanged.connect(self._validate)
        self.select_loader(Loader.FABRIC)
        self._validate()

    def _field_label(self, text: str, parent: QWidget) -> QLabel:
        label = QLabel(f"<span style='color:#00e5cc'>&gt;</span> {text}", parent)
        label.setStyleSheet("font-size: 9px; color: #7ab8d4;")
        return label

    def _add_field(self, body: QWidget, body_layout: QVBoxLayout, label: str, placeholder: str) -> QLineEdit:
        block = QWidget(body)
        layout = QVBoxLayout(block)
        layout.setContentsMargins(0, 0, 0, 0)
        layout.setSpacing(5)
        layout.addWidget(self._field_label(label, block))
        edit = QLineEdit(block)
        edit.setObjectName("FieldInput")
        edit.setPlaceholderText(placeholder)
        layout.addWidget(edit)
        body_layout.addWidget(block)
        return edit

    def _make_loader_tile(self, parent: QWidget, abbr: str, label: str, loader: Loader) -> QFrame:
        tile = QFrame(parent)
        tile.setFixedHeight(56)
        tile.setCursor(Qt.PointingHandCursor)
        layout = QVBoxLayout(tile)
        layout.setContentsMargins(0, 6, 0, 6)
        layout.setAlignment(Qt.AlignCenter)
        abbr_label = QLabel(abbr, tile)
        abbr_label.setObjectName("LoaderAbbr")
        abbr_label.setAlignment(Qt.AlignCenter)
        name_label = QLabel(label, tile)
        name_label.setAlignment(Qt.AlignCenter)
        name_label.setStyleSheet("font-size:8px; color:#4a7a96; text-transform:uppercase;")
        layout.addWidget(abbr_label)
        layout.addWidget(name_label)
        tile.mousePressEvent = lambda _event, l=loader: self.select_loader(l)
        return tile

    def select_loader(self, loader: Loader) -> None:
        self.loader = loader
        for candidate, tile in self.loader_tiles.items():
            selected = candidate == loader
            tile.setStyleSheet(
                "QFrame { background: qlineargradient(x1:0,y1:0,x2:0.5,y2:1,"
                + (
                    "stop:0 rgba(56,217,245,0.12),stop:1 rgba(20,80,110,0.15));"
                    "border-top:1px solid rgba(56,217,245,0.5);"
                    "border-left:1px solid rgba(56,217,245,0.22);"
                    if selected
                    else "stop:0 rgba(10,30,52,0.7),stop:1 rgba(4,14,26,0.8));"
                    "border-top:1px solid rgba(255,255,255,0.08);"
                    "border-left:1px solid rgba(255,255,255,0.04);"
                )
                + "border-right:1px solid rgba(0,0,0,0.25);"
                "border-bottom:1px solid rgba(0,0,0,0.45);"
                "border-radius:6px; }"
            )
            labels = tile.findChildren(QLabel)
            if labels:
                labels[0].setStyleSheet(
                    "font-family:'Exo 2'; font-weight:900; font-size:12px;"
                    f"color:{'#38d9f5' if selected else '#7ab8d4'};"
                )

    def _validate(self) -> None:
        ok = bool(self.name_edit.text().strip()) and bool(self.version_edit.text().strip())
        self.create_button.setEnabled(ok)
        self.create_button.setObjectName("BtnPrimary" if ok else "BtnSecondary")
        self.create_button.style().unpolish(self.create_button)
        self.create_button.style().polish(self.create_button)

    def _accept_result(self) -> None:
        self.result_data = NewPackData(
            name=self.name_edit.text().strip(),
            version=self.version_edit.text().strip(),
            loader=self.loader,
        )
        self.accept()
